File: app/src/main/java/cn/guluwa/gulumusic/ui/viewmodel/MainViewModel.java
package cn.guluwa.gulumusic.ui.viewmodel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.view.View;

import java.io.File;
import java.util.List;

import cn.guluwa.gulumusic.data.bean.FreshBean;
import cn.guluwa.gulumusic.data.bean.LocalSongBean;
import cn.guluwa.gulumusic.data.bean.SearchResultSongBean;
import cn.guluwa.gulumusic.data.bean.SongPathBean;
import cn.guluwa.gulumusic.data.bean.SongWordBean;
import cn.guluwa.gulumusic.data.bean.TracksBean;
import cn.guluwa.gulumusic.data.bean.ViewDataBean;
import cn.guluwa.gulumusic.data.total.SongsRepository;
import cn.guluwa.gulumusic.listener.OnResultListener;

/**
 * Created by guluwa on 2018/1/12.
 */

public class MainViewModel extends ViewModel {

    private SongsRepository songsRepository = SongsRepository.getInstance();

    //热门歌曲刷新
    private MutableLiveData<FreshBean> mHotSongListFresh;
    //热门歌曲
    private LiveData<ViewDataBean<List<TracksBean>>> mHotSongs;

    /**
     * 查询热门歌曲
     *
     * @return
     */
    public LiveData<ViewDataBean<List<TracksBean>>> queryNetCloudHotSong() {
        if (mHotSongs == null) {
            mHotSongListFresh = new MutableLiveData<>();
            mHotSongs = Transformations.switchMap(mHotSongListFresh, input -> {
                if (input.isFresh) {
                    return songsRepository.queryNetCloudHotSong(input.isFirstComing);
                } else {
                    return null;
                }
            });
        }
        return mHotSongs;
    }

    /**
     * 刷新热门歌曲
     *
     * @param isFresh
     * @param isFirstComing
     */
    public void refreshHot(boolean isFresh, boolean isFirstComing) {
        mHotSongListFresh.setValue(new FreshBean(isFresh, isFirstComing));
    }

    //本地歌曲刷新
    private MutableLiveData<Boolean> mLocalSongListFresh;
    //本地歌曲
    private LiveData<ViewDataBean<List<LocalSongBean>>> mLocalSongs;

    /**
     * 查询本地歌曲
     *
     * @return
     */
    public LiveData<ViewDataBean<List<LocalSongBean>>> queryLocalSong() {
        if (mLocalSongs == null) {
            mLocalSongListFresh = new MutableLiveData<>();
            mLocalSongs = Transformations.switchMap(mLocalSongListFresh, input -> {
                if (input) {
                    return songsRepository.queryLocalSong();
                } else {
                    return null;
                }
            });
        }
        return mLocalSongs;
    }

    /**
     * 刷新本地歌曲
     *
     * @param isFresh
     */
    public void refreshLocal(boolean isFresh) {
        mLocalSongListFresh.setValue(isFresh);
    }

    //歌曲搜索
    private MutableLiveData<FreshBean> mSearchSongListFresh;
    //搜索歌曲
    private LiveData<ViewDataBean<List<SearchResultSongBean>>> mSearchSongs;

    /**
     * 歌曲搜索
     *
     * @return
     */
    public LiveData<ViewDataBean<List<SearchResultSongBean>>> searchSongByKeyWord() {
        if (mSearchSongs == null) {
            mSearchSongListFresh = new MutableLiveData<>();
            mSearchSongs = Transformations.switchMap(mSearchSongListFresh, input -> {
                if (input.page != -1) {
                    return songsRepository.searchSongByKeyWord(input);
                } else {
                    return null;
                }
            });
        }
        return mSearchSongs;
    }

    /**
     * 刷新歌曲搜索
     */
    public void refreshSearchSongs(String key, int page) {
        mSearchSongListFresh.setValue(new FreshBean(key, page));
    }
}

File: app/src/main/java/cn/guluwa/gulumusic/ui/viewmodel/MainViewModel.kt
package cn.guluwa.gulumusic.ui.viewmodel

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import android.arch.lifecycle.ViewModel

import cn.guluwa.gulumusic.data.bean.FreshBean
import cn.guluwa.gulumusic.data.bean.LocalSongBean
import cn.guluwa.gulumusic.data.bean.SearchResultSongBean
import cn.guluwa.gulumusic.data.bean.TracksBean
import cn.guluwa.gulumusic.data.bean.ViewDataBean
import cn.guluwa.gulumusic.data.total.SongsRepository

/**
 * Created by guluwa on 2018/1/12.
 */

class MainViewModel : ViewModel() {

    //热门歌曲刷新
    private var mHotSongListFresh: MutableLiveData<FreshBean>? = null
    //热门歌曲
    private var mHotSongs: LiveData<ViewDataBean<List<TracksBean>>>? = null

    //本地歌曲刷新
    private var mLocalSongListFresh: MutableLiveData<Boolean>? = null
    //本地歌曲
    private var mLocalSongs: LiveData<ViewDataBean<List<LocalSongBean>>>? = null

    //歌曲搜索
    private var mSearchSongListFresh: MutableLiveData<FreshBean>? = null
    //搜索歌曲
    private var mSearchSongs: LiveData<ViewDataBean<List<SearchResultSongBean>>>? = null

    /**
     * 查询热门歌曲
     *
     * @return
     */
    fun queryNetCloudHotSong(): LiveData<ViewDataBean<List<TracksBean>>>? {
        if (mHotSongs == null) {
            mHotSongListFresh = MutableLiveData()
            mHotSongs = Transformations.switchMap(mHotSongListFresh!!) { input ->
                if (input.isFresh) {
                    SongsRepository.getInstance().queryNetCloudHotSong(input.isFirstComing)
                } else {
                    null
                }
            }
        }
        return mHotSongs
    }

    /**
     * 刷新热门歌曲
     *
     * @param isFresh
     * @param isFirstComing
     */
    fun refreshHot(isFresh: Boolean, isFirstComing: Boolean) {
        mHotSongListFresh!!.value = FreshBean(isFresh, isFirstComing)
    }

    /**
     * 查询本地歌曲
     *
     * @return
     */
    fun queryLocalSong(): LiveData<ViewDataBean<List<LocalSongBean>>>? {
        if (mLocalSongs == null) {
            mLocalSongListFresh = MutableLiveData()
            mLocalSongs = Transformations.switchMap(mLocalSongListFresh!!) { input ->
                if (input!!) {
                    SongsRepository.getInstance().queryLocalSong()
                } else {
                    null
                }
            }
        }
        return mLocalSongs
    }

    /**
     * 刷新本地歌曲
     *
     * @param isFresh
     */
    fun refreshLocal(isFresh: Boolean) {
        mLocalSongListFresh!!.value = isFresh
    }

    /**
     * 歌曲搜索
     *
     * @return
     */
    fun searchSongByKeyWord(): LiveData<ViewDataBean<List<SearchResultSongBean>>>? {
        if (mSearchSongs == null) {
            mSearchSongListFresh = MutableLiveData()
            mSearchSongs = Transformations.switchMap(mSearchSongListFresh!!) { input ->
                if (input.isFresh) {
                    SongsRepository.getInstance().searchSongByKeyWord(input)
                } else {
                    null
                }
            }
        }
        return mSearchSongs
    }

    /**
     * 刷新歌曲搜索
     */
    fun refreshSearchSongs(key: String, page: Int, isFresh: Boolean) {
        mSearchSongListFresh!!.value = FreshBean(key, page, isFresh)
    }
}
